module-system/sov-state/src/vec.rs

```diff
@@ -0,0 +1,201 @@
+use thiserror::Error;
+
+use crate::codec::{BorshCodec, StateKeyCodec, StateValueCodec};
+use crate::{Prefix, Storage, WorkingSet};
+
+#[derive(Debug, PartialEq, Eq, Clone, BorshDeserialize, BorshSerialize)]
+pub struct StateVec<V, C = BorshCodec> {
+    _phantom: PhantomData<V>,
+    codec: C,
+    prefix: Prefix,
+}
+
+/// Error type for `StateVec` get method.
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Index out of bounds for index: {0}")]
+    IndexOutOfBounds(usize),
+    #[error("Value not found for prefix: {0} and index: {1}")]
+    MissingValue(Prefix, usize),
+}
+
+impl<V> StateVec<V>
+where
+    BorshCodec: StateValueCodec<V>,
+{
+    /// Crates a new [`StateVec`] with the given prefix and the default
+    /// [`StateCodec`] (i.e. [`BorshCodec`]).
+    pub fn new(prefix: Prefix) -> Self {
+        Self {
+            _phantom: PhantomData,
+            codec: BorshCodec,
+            prefix,
+        }
+    }
+}
+
+impl<V, C> StateVec<V, C>
+where
+    C: StateValueCodec<V>,
+    C: StateValueCodec<usize>,
+{
+    /// Creates a new [`StateVec`] with the given prefix and codec.
+    pub fn with_codec(prefix: Prefix, codec: C) -> Self {
+        Self {
+            _phantom: PhantomData,
+            codec,
+            prefix,
+        }
+    }
+
+    /// Returns the prefix used when this [`StateVec`] was created.
+    pub fn prefix(&self) -> &Prefix {
+        &self.prefix
+    }
+
+    fn internal_codec(&self) -> IndexCodec<C> {
+        IndexCodec::new(&self.codec)
+    }
+
+    /// Sets a value in the [`StateVec`].
+    /// If the index is out of bounds, returns an error.
+    /// To push a value to the end of the StateVec, use [`StateVec::push`].
+    pub fn set<S: Storage>(&self, index: usize, value: &V, working_set: &mut WorkingSet<S>) -> Result<(), Error> {
+        let len = self.len(working_set);
+
+        if index < len {
+            working_set.set_value(self.prefix(), &self.internal_codec(), &IndexKey(index + 1), value);
+            Ok(())
+        } else {
+            Err(Error::IndexOutOfBounds(index))
+        }
+    }
+
+    /// Returns the value for the given index.
+    pub fn get<S: Storage>(&self, index: usize, working_set: &mut WorkingSet<S>) -> Option<V> {
+        working_set.get_value(self.prefix(), &self.internal_codec(), &IndexKey(index + 1))
+    }
+
+    /// Returns the value for the given index.
+    /// If the index is out of bounds, returns an error.
+    /// If the value is absent, returns an error.
+    pub fn get_or_err<S: Storage>(&self, index: usize, working_set: &mut WorkingSet<S>) -> Result<Option<V>, Error> {
+        let len = self.len(working_set);
+
+        if index < len {
+            let elem = working_set.get_value(self.prefix(), &self.internal_codec(), &IndexKey(index + 1));
+
+            if elem.is_some() {
+                Ok(elem)
+            } else {
+                Err(Error::MissingValue(self.prefix().clone(), index))
+            }
+        } else {
```

Comment:
How it can be None? Does unitialized value return none or 0?

Reply:
underlying storage mechanism returns none for value never set